File: hackerrank/The Longest Increasing Subsequence.py
def smallest_greater_dc(array, a, b, value):
    while a != b:
        c = (a + b) // 2
        if array[c] < value:
            a = c + 1
        else:
            b = c
    return b

def longest_increasing_subsequence(sequence):
    dp = [sequence[0]]
    for i in range(1, len(sequence)):
        if sequence[i] > dp[-1]:
            dp.append(sequence[i])
        elif sequence[i] < dp[-1]:
            dp[smallest_greater_dc(dp, 0, len(dp) - 1, sequence[i])] = sequence[i]
    return len(dp)

if __name__ == '__main__':
    sequence = [int(input()) for i in range(int(input()))]
    print(longest_increasing_subsequence(sequence))

File: hackerrank/The Longest Increasing Subsequence.c
#include<stdio.h>
#define MAX_N 1000000

int sequence[MAX_N], dp[MAX_N];

int read_test_case(int sequence[]){
    int i, n;
    scanf("%d", &n);
    for(i = 0; i < n; i++)
        scanf("%d\n", &sequence[i]);
    return n;
}

int smallest_greater_dc(int array[], int a, int b, int value){
    int c;
    while(a != b){
        c = (a + b) / 2;
        if(array[c] < value)
            a = c + 1;
        else 
            b = c;
    }
    return b;
}

int longest_increasing_subsequence(int n, int sequence[], int dp[]){
    int i, dp_n = 1;
    dp[0] = sequence[0];
    for(i = 1; i < n; i++){
        if(sequence[i] > dp[dp_n - 1]){
            dp[dp_n++] = sequence[i];
        }else if(sequence[i] < dp[dp_n - 1]){
            dp[smallest_greater_dc(dp, 0, dp_n - 1, sequence[i])] = sequence[i];
        }
    }
    return dp_n;
}

int main(){
    int n;
    n = read_test_case(sequence);
    printf("%d\n", longest_increasing_subsequence(n, sequence, dp));
    return 0;
}
